test/unit/puppetlabs/services/jruby_pool_manager/jruby_ref_pool_test.clj

```diff
@@ -1,2 +1,107 @@
 (ns puppetlabs.services.jruby-pool-manager.jruby-ref-pool-test
-  (:require [clojure.test :refer :all]))
+  (:require [clojure.test :refer :all]
+            [puppetlabs.services.jruby-pool-manager.jruby-testutils :as jruby-testutils]
+            [puppetlabs.services.jruby-pool-manager.jruby-core :as jruby-core]
+            [puppetlabs.services.protocols.jruby-pool :as pool-protocol]))
+
+(defn jruby-test-config
+  ([max-borrows]
+   (jruby-test-config max-borrows 1))
+  ([max-borrows max-instances]
+   (jruby-testutils/jruby-config {:max-active-instances max-instances
+                                  :multithreaded true
+                                  :max-borrows-per-instance max-borrows})))
+
+(deftest more-borrows-than-mrpi
+  (testing "flushing due to max borrows succeeds when we've over-borrowed")
+  (jruby-testutils/with-pool-context
+    pool-context
+    jruby-testutils/default-services
+    ;; We can check out the instance more times than `max-borrows` and return
+    ;; them each in sequence, and nothing blocks or fails
+    (jruby-testutils/jruby-config {:max-active-instances 4
+                                   :multithreaded true
+                                   :max-borrows-per-instance 3})
+    (let [instance (jruby-core/borrow-from-pool pool-context :test [])]
+      (is (= 1 (:id instance)))
+      (jruby-core/return-to-pool pool-context instance :test []))
+    (jruby-testutils/drain-and-refill pool-context)
+    (let [instance (jruby-core/borrow-from-pool pool-context :test [])]
+      (is (= 2 (:id instance)))
+      (jruby-core/return-to-pool pool-context instance :test []))))
+
+
+(deftest flush-jruby-after-max-borrows
+  (testing "JRubyInstance is not flushed if it has not exceeded max borrows"
+    (jruby-testutils/with-pool-context
+      pool-context
+      jruby-testutils/default-services
+      (jruby-test-config 2)
+      (let [instance (jruby-core/borrow-from-pool pool-context :test [])
+            id (:id instance)]
+        (jruby-core/return-to-pool pool-context instance :test [])
+        (let [instance (jruby-core/borrow-from-pool pool-context :test [])]
+          (is (= id (:id instance)))
+          (jruby-core/return-to-pool pool-context instance :test [])))))
+  (testing "JRubyInstance is flushed after exceeding max borrows"
+    (jruby-testutils/with-pool-context
+      pool-context
+      jruby-testutils/default-services
+      (jruby-test-config 2)
+      (jruby-testutils/wait-for-jrubies-from-pool-context pool-context)
```

Comment:
I'm a little confused as to why this is used here and nowhere else.

Reply:
This test wound up getting deleted, but I'm thinking maybe it shouldn't have.

Reply:
Just kidding, we're still functionally testing the same scenario without it.